Hint suggestion: suggestion on failing to download `https://github.com/foo/bar`

```
$ opam pin add ocamlbuild https://github.com/ocaml/ocamlbuild
[ocamlbuild.0.12.0] downloaded from https://github.com/ocaml/ocamlbuild
[ERROR] Error getting source from https://github.com/ocaml/ocamlbuild
```

It could be nice to print a hint in this case, that one should give the URL of the git directory (as used for `git clone`) rather than the URL of the website. Maybe just, on all such "error getting source from", a small explanation of the various accepted URL shapes could be given.

(At first I tried to give a regexp to capture github and gitlab etc., but I couldn't find a nice heuristic for this.)

--- a/src/client/opamAction.ml
+++ b/src/client/opamAction.ml
@@ -347,17 +347,21 @@ let get_wrappers t =
     ~inner:(OpamFile.Switch_config.wrappers t.switch_config)
 
 let get_wrapper t opam wrappers ?local getter =
-  OpamFilter.commands (OpamPackageVar.resolve ?local ~opam t)
+  let root = t.switch_global.root in
+  let hook_vnam = OpamVariable.of_string "hooks" in
+  let hook_vval = Some (OpamVariable.dirname (OpamPath.hooks_dir root)) in
+  let local_env = match local with
+    | Some e -> OpamVariable.Map.add hook_vnam hook_vval e
+    | None ->OpamVariable.Map.singleton hook_vnam hook_vval
+  in
+  OpamFilter.commands (OpamPackageVar.resolve ~local:local_env ~opam t)
     (getter wrappers) |>
   OpamStd.List.filter_map (function
       | [] -> None
       | cmd::args -> Some (cmd, args))
 
 let cmd_wrapper t opam wrappers getter cmd args =
-  let root = t.switch_global.root in
-  let local = OpamVariable.Map.singleton (OpamVariable.of_string "hooks")
-      (Some (OpamVariable.dirname (OpamPath.hooks_dir root))) in
-  match get_wrapper t opam wrappers ~local getter @ [cmd, args] with
+  match get_wrapper t opam wrappers getter @ [cmd, args] with
   | (cmd, args) :: r -> cmd, args @ List.concat (List.map (fun (c, a) -> c::a) r)
   | [] -> assert false
 

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -459,7 +459,7 @@ let update
          package%s with %s %s"
         (OpamPackage.Set.to_string nondev_packages)
         (OpamConsole.colorise `bold "opam update")
-        (if OpamPackage.Set.cardinal nondev_packages = 1 then "" else "s")
+        (if OpamPackage.Set.is_singleton nondev_packages then "" else "s")
         (OpamConsole.colorise `bold "opam upgrade")
         (OpamConsole.colorise `bold
            (OpamStd.List.concat_map " " OpamPackage.name_to_string

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -613,37 +613,45 @@ let init_checks ?(hard_fail_exn=true) config =
   if hard_fail && hard_fail_exn then OpamStd.Sys.exit_because `Configuration_error
   else not (soft_fail || hard_fail)
 
-let update_with_init_config ?(overwrite=false) config init_config =
+let update_with_init_config_without_scripts, update_with_init_config =
   let module I = OpamFile.InitConfig in
   let module C = OpamFile.Config in
-  let setifnew getter setter v conf =
+  let setifnew overwrite getter setter v conf =
     if overwrite then setter v conf
     else if getter conf = getter C.empty then setter v conf
     else conf
   in
-  config |>
-  setifnew C.jobs C.with_jobs (match I.jobs init_config with
-      | Some j -> j
-      | None -> Lazy.force OpamStateConfig.(default.jobs)) |>
-  setifnew C.dl_tool C.with_dl_tool_opt (I.dl_tool init_config) |>
-  setifnew C.dl_jobs C.with_dl_jobs
-    (OpamStd.Option.default OpamStateConfig.(default.dl_jobs)
-       (I.dl_jobs init_config)) |>
-  setifnew C.criteria C.with_criteria (I.solver_criteria init_config) |>
-  setifnew C.solver C.with_solver_opt (I.solver init_config) |>
-  setifnew C.wrappers C.with_wrappers (I.wrappers init_config) |>
-  setifnew C.global_variables C.with_global_variables
-    (I.global_variables init_config) |>
-  setifnew C.eval_variables C.with_eval_variables
-    (I.eval_variables init_config) |>
-  setifnew C.default_compiler C.with_default_compiler
-    (I.default_compiler init_config) |>
-  setifnew C.recommended_tools C.with_recommended_tools
-    (I.recommended_tools init_config) |>
-  setifnew C.required_tools C.with_required_tools
-    (I.required_tools init_config) |>
-  setifnew C.init_scripts C.with_init_scripts
-    (I.init_scripts init_config)
+  let write ?(overwrite=false) config init_config =
+    let setifnew g = setifnew overwrite g in
+    config |>
+    setifnew C.jobs C.with_jobs (match I.jobs init_config with
+        | Some j -> j
+        | None -> Lazy.force OpamStateConfig.(default.jobs)) |>
+    setifnew C.dl_tool C.with_dl_tool_opt (I.dl_tool init_config) |>
+    setifnew C.dl_jobs C.with_dl_jobs
+      (OpamStd.Option.default OpamStateConfig.(default.dl_jobs)
+         (I.dl_jobs init_config)) |>
+    setifnew C.criteria C.with_criteria (I.solver_criteria init_config) |>
+    setifnew C.solver C.with_solver_opt (I.solver init_config) |>
+    setifnew C.wrappers C.with_wrappers (I.wrappers init_config) |>
+    setifnew C.global_variables C.with_global_variables
+      (I.global_variables init_config) |>
+    setifnew C.eval_variables C.with_eval_variables
+      (I.eval_variables init_config) |>
+    setifnew C.default_compiler C.with_default_compiler
+      (I.default_compiler init_config)
+  in
+  let full ?(overwrite=false) config init_config =
+    let setifnew g = setifnew overwrite g in
+    config |>
+    setifnew C.recommended_tools C.with_recommended_tools
+      (I.recommended_tools init_config) |>
+    setifnew C.required_tools C.with_required_tools
+      (I.required_tools init_config) |>
+    setifnew C.init_scripts C.with_init_scripts
+      (I.init_scripts init_config)
+  in
+  write,full
 
 let reinit ?(init_config=OpamInitDefaults.init_config()) config =
   let root = OpamStateConfig.(!r.root_dir) in
@@ -699,12 +707,19 @@ let init
           | Some r -> [r.repo_name, (r.repo_url, r.repo_trust)]
           | None -> OpamFile.InitConfig.repositories init_config
         in
-        let config =
-          update_with_init_config OpamFile.Config.empty init_config |>
+        let config_repo =
           OpamFile.Config.with_repositories (List.map fst repos)
         in
-        OpamFile.Config.write config_f config;
+        let w_config =
+          update_with_init_config_without_scripts OpamFile.Config.empty
+            init_config |> config_repo
+        in
+        OpamFile.Config.write config_f w_config;
 
+        let config =
+          update_with_init_config w_config init_config |>
+          config_repo
+        in
         let dontswitch =
           if bypass_checks then false else
           let all_ok = init_checks config in

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -673,7 +673,7 @@ let reinit ?(init_config=OpamInitDefaults.init_config()) config =
   gt, rt
 
 let init
-    ?(init_config=OpamInitDefaults.init_config()) ?(show_opamrc=false)
+    ?(init_config=OpamInitDefaults.init_config())
     ?repo ?(bypass_checks=false) shell dot_profile update_config =
   log "INIT %a"
     (slog @@ OpamStd.Option.to_string OpamRepositoryBackend.to_string) repo;
@@ -703,11 +703,6 @@ let init
           update_with_init_config OpamFile.Config.empty init_config |>
           OpamFile.Config.with_repositories (List.map fst repos)
         in
-        (* If show option is set, dump it and exit *)
-        if show_opamrc then
-          (OpamFile.InitConfig.write_to_channel stdout init_config;
-           OpamStd.Sys.exit_because `Success);
-        (* Else write the config file and continue init *)
         OpamFile.Config.write config_f config;
 
         let dontswitch =

--- a/src/client/opamClient.mli
+++ b/src/client/opamClient.mli
@@ -18,7 +18,6 @@ open OpamStateTypes
 (** Initialize the client to a consistent state. *)
 val init:
   ?init_config:OpamFile.InitConfig.t ->
-  ?show_opamrc:bool ->
   ?repo:repository ->
   ?bypass_checks:bool ->
   shell -> filename -> [`ask|`yes|`no] ->

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -2265,7 +2265,12 @@ let pin ?(unpin_only=false) () =
         (OpamPinned.files_in_source dir)
     in
     let basename =
-      List.hd (OpamStd.String.split (OpamUrl.basename url) '.')
+      match OpamStd.String.split (OpamUrl.basename url) '.' with
+      | [] ->
+        OpamConsole.error_and_exit `Bad_arguments
+          "Can not retrieve a path from '%s'"
+          (OpamUrl.to_string url)
+      | b::_ -> b
     in
     let found =
       match OpamUrl.local_dir url with

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -2415,7 +2415,7 @@ let pin ?(unpin_only=false) () =
          let st =
            List.fold_left (fun st name ->
                try OpamPinCommand.source_pin st name ~edit (Some url)
-               with OpamPinCommand.Aborted
+               with OpamPinCommand.Aborted -> OpamStd.Sys.exit_because `Aborted
                   | OpamPinCommand.Nothing_to_do -> st)
              st names
          in

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -1830,7 +1830,8 @@ let switch =
     "Set the currently active switch, among the installed switches.";
     "remove", `remove, ["SWITCH"], "Remove the given switch from disk.";
     "export", `export, ["FILE"],
-    "Save the current switch state to a file.";
+    "Save the current switch state to a file. If $(b,--full) is specified, it \
+     includes the metadata of all installed packages";
     "import", `import, ["FILE"],
     "Import a saved switch state. If $(b,--switch) is specified and doesn't \
      point to an existing switch, the switch will be created for the import.";
@@ -1840,8 +1841,8 @@ let switch =
     "Lists installed switches.";
     "list-available", `list_available, ["[PATTERN]"],
     "Lists base packages that can be used to create a new switch, i.e. \
-     packages with the $(i,compiler) flag set. Only standard versions are \
-     shown by default if no pattern is supplied, use $(b,--all) to show all.";
+     packages with the $(i,compiler) flag set. If no pattern is supplied, \
+     all versions are shown.";
     "show", `current, [], "Prints the name of the current switch.";
     "set-base", `set_compiler, ["PACKAGES"],
     "Sets the packages forming the immutable base for the selected switch, \
@@ -1883,7 +1884,7 @@ let switch =
   let command, params = mk_subcommands_with_default commands in
   let no_switch =
     mk_flag ["no-switch"]
-      "Don't automatically select newly installed switches" in
+      "Don't automatically select newly installed switches." in
   let packages =
     mk_opt ["packages"] "PACKAGES"
       "When installing a switch, explicitely define the set of packages to set \
@@ -1915,7 +1916,7 @@ let switch =
     mk_flag ["full"]
       "When exporting, include the metadata of all installed packages, \
        allowing to re-import even if they don't exist in the repositories (the \
-       default is to include only the metadata of pinned packages)"
+       default is to include only the metadata of pinned packages)."
   in
   let no_install =
     mk_flag ["no-install"]

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -244,6 +244,11 @@ let init =
         OpamConsole.errmsg "%s\n" (Printexc.to_string e);
         OpamStd.Sys.exit_because `Configuration_error
     in
+    (* If show option is set, dump it and exit *)
+    if show_opamrc then
+      (OpamFile.InitConfig.write_to_channel stdout init_config;
+       OpamStd.Sys.exit_because `Success);
+    (* Else continue init *)
     let repo =
       OpamStd.Option.map (fun url ->
         let repo_url = OpamUrl.parse ?backend:repo_kind url in
@@ -261,7 +266,7 @@ let init =
     let dot_profile = init_dot_profile shell dot_profile_o in
     let gt, rt, default_compiler =
       OpamClient.init
-        ~init_config ~show_opamrc ?repo ~bypass_checks
+        ~init_config ?repo ~bypass_checks
         shell dot_profile update_config
     in
     if not no_compiler &&

--- a/src/client/opamPinCommand.ml
+++ b/src/client/opamPinCommand.ml
@@ -50,6 +50,8 @@ let read_opam_file_for_pinning ?(quiet=false) name f url =
     in
     Some opam
 
+exception Fetch_Fail of string
+
 let get_source_definition ?version st nv url =
   let root = st.switch_global.root in
   let srcdir = OpamPath.Switch.pinned_package root st.switch nv.name in
@@ -62,7 +64,7 @@ let get_source_definition ?version st nv url =
   in
   let open OpamProcess.Job.Op in
   OpamUpdate.fetch_dev_package url srcdir nv @@| function
-  | Not_available _ -> raise Not_found
+  | Not_available s -> raise (Fetch_Fail s)
   | Up_to_date _ | Result _ ->
     match OpamPinned.find_opam_file_in_source nv.name srcdir with
     | None -> None
@@ -449,11 +451,12 @@ and source_pin
         urlf >>= fun url ->
         OpamProcess.Job.run @@ get_source_definition ?version st nv url
       )
-    with Not_found ->
+    with Fetch_Fail err ->
       if force then None else
-        OpamConsole.error_and_exit `Sync_error
-          "Error getting source from %s"
-          (OpamStd.Option.to_string OpamUrl.to_string target_url)
+        (OpamConsole.error_and_exit `Sync_error
+           "Error getting source from %s:\n%s"
+           (OpamStd.Option.to_string OpamUrl.to_string target_url)
+           (OpamStd.Format.itemize (fun x -> x) [err]));
   in
 
   let nv =

